R/interactive_plots.R

```diff
@@ -0,0 +1,244 @@
+#' @param sn Optional numeric value specifying the signal-to-noise ratio threshold. If provided along with \code{min_sn}, regions with SNR below the threshold will be excluded from the heatmap.
+#' @param cor Optional numeric value specifying the correlation threshold. If provided along with \code{min_cor}, regions with correlation below the threshold will be excluded from the heatmap.
+#' @param min_sn Optional numeric value specifying the minimum signal-to-noise ratio for inclusion in the heatmap. Regions with SNR below this threshold will be excluded.
+#' @param min_cor Optional numeric value specifying the minimum correlation for inclusion in the heatmap. Regions with correlation below this threshold will be excluded.
+#' @param selected_spectrum Optional index of the selected spectrum to highlight on the heatmap.
+#' @param selected_spectrum2 an optional second index of the selected spectrum to highlight on the heatmap.
+#' @param \ldots further arguments passed to subfunctions.
+#'
+#' @return A plotly heatmap object displaying the OpenSpecy data. A subplot containing the heatmap and spectra plot. A plotly object displaying the spectra from the \code{OpenSpecy} object(s).
+#'
+#' @examples
+#' data("raman_hdpe")
+#' tiny_map <- read_zip(read_extdata("CA_tiny_map.zip"))
+#' plotly_spec(raman_hdpe)
+#'
+#' correlation <- cor_spec(
+#'   conform_spec(raman_hdpe, new_wavenumbers = raman_hdpe$wavenumber, res = 5),
+#'   conform_spec(tiny_map, tiny_map$wavenumbers, res = 5)
+#' )
+#' heatmap_spec(tiny_map, z = tiny_map$metadata$y)
+#'
+#' sample_spec(tiny_map, size = 12) |>
+#'   interactive_plot(selected_spectrum = 2, x2 = raman_hdpe)
+#'
+#' @author
+#' Win Cowger, Zacharias Steinmetz
+#'
+#' @importFrom plotly plot_ly add_trace add_markers subplot layout
+#' @importFrom data.table melt
+#'
+#' @export
+plotly_spec <- function(x, ...) {
+  UseMethod("plotly_spec")
+}
+
+#' @rdname interactive_plots
+#'
+#' @export
+plotly_spec.default <- function(x, ...) {
+  stop("object 'x' needs to be of class 'OpenSpecy'")
+}
+
+#' @rdname interactive_plots
+#'
+#' @export
+plotly_spec.OpenSpecy <- function(x, x2 = NULL, ...) {
+  dt <- cbind(wavenumber = x$wavenumber, x$spectra) |>
+    melt(id.vars = "wavenumber", variable.name = "id", value.name = "intensity")
+
+  p <- plot_ly(dt, type = "scatter", mode = "lines", ...) |>
+    add_trace(x = ~wavenumber,
+              y = ~make_rel(intensity, na.rm = T),
+              color = ~id,
+              name = "Your Spectra",
+              line = list(color = 'rgb(255,255,255)'),
+              showlegend = F) |>
+    layout(xaxis = list(title = "wavenumber [cm<sup>-1</sup>]", autorange = "reversed"),
+           yaxis = list(title = "absorbance intensity [-]"),
+           plot_bgcolor = 'rgb(17,0,73)',
+           paper_bgcolor = 'rgb(0,0,0)',
+           legend = list(orientation = 'h', y = 1.1),
+           font = list(color = '#FFFFFF'))
```

Comment:
I'd opt for making those functions more generic without presetting a certain style. We can still pass those layout parameters as function arguments in the Shiny app. What do you think?